[librustdoc] Only split lang string on `,`, ` `, and `\t`

casey opened the pull request:
Split markdown lang strings into tokens on `,`.

The previous behavior was to split lang strings into tokens on any
character that wasn't a `_`, `_`, or alphanumeric.

This is a potentially breaking change, so please scrutinize! See discussion in #78344.

I noticed some test cases that made me wonder if there might have been some reason for the original behavior:

```
t("{.no_run .example}", false, true, Ignore::None, true, false, false, false, v(), None);
t("{.sh .should_panic}", true, false, Ignore::None, false, false, false, false, v(), None);
t("{.example .rust}", false, false, Ignore::None, true, false, false, false, v(), None);
t("{.test_harness .rust}", false, false, Ignore::None, true, true, false, false, v(), None);
```

It seemed pretty peculiar to specifically test lang strings in braces, with all the tokens prefixed by `.`.

I did some digging, and it looks like the test cases were added way back in [this commit from 2014](https://github.com/rust-lang/rust/commit/3fef7a74ca9a) by @skade.

It looks like they were added just to make sure that the splitting was permissive, and aren't testing that those strings in particular are accepted.

Closes https://github.com/rust-lang/rust/issues/78344.

rust-highfive commented:
r? @GuillaumeGomez

(rust_highfive has picked a reviewer for you, use r? to override)

jyn514 commented:
@rfcbot fcp merge

This changes the parsing of language strings for code blocks so that only `,` is accepted as a delimiter. Previously, any character that couldn't be part of an identifier was treated as a delimiter. This gives rustdoc increased flexibility to distinguish betweeen delimiters, such as for https://github.com/casey/rfcs/blob/doctest-name/text/0000-doctest-name.md.

This is a breaking change, but in practice it seems unlikely anyone is using this. It's possible that we could find out with a crater run if crater allows denying specific warnings.

Possible alternatives:
- Only disallow `=`. This still unblocks the extension @casey proposes in the linked RFC, but allows using all other non-identifiers as delimiters. Any future extensions to lang strings will have to reserve more delimiters at that time.
- Keep the status quo. This is not a breaking change, but means that rustdoc will never be able to add more extensions to language strings.

jyn514 commented:
@casey as-is, this breaks trailing text after spaces:

```
/// ignore (tidy)
/// goo();
/// ```
pub fn bar() -> usize { 2 }
```
```
6: @has check failed
	`XPATH PATTERN` did not match
	// @has foo/fn.bar.html '//*[@class="tooltip ignore"]/span' "This example is not tested"
```
What do you think about changing this to split on either spaces or `,` instead?

casey commented:
@jyn514 Sounds good to me! I think using spaces is common, particularly with GitHub flavored markdown, so splitting on spaces seems reasonable. I also split on tabs, since it seems weird to split on spaces but not tabs, and since users sometimes use tabs for alignment.

GuillaumeGomez commented:
Hum also: we use such headers in rust tests:

``````
```rust,ignore (whatever is the reason)
``````

So splitting on whistepaces and tabs might not be a good idea...

casey commented:
Unrecognized tokens are ignored, so this shouldn't change the set of recognized tokens in the above example.

Before the change, `rust,ignore (whatever is the reason)` will be split into `rust`, `ignore`, `whatever`, `is`, `the`, `reason`, and all but `rust` and `ignore` are ignored.

After the change, `rust,ignore (whatever is the reason)` will be split into `rust`, `ignore`, `(whatever`, `is`, `the`, `reason)`, and all but `rust` and `ignore` are ignored.

GuillaumeGomez commented:
Yes but what's the point of splitting on whitespaces and tabs?

casey commented:
If it didn't split on spaces, then something like `ignore (tidy)` will be split as `ignore (tidy)` (one token) and so won't work. This seems to be somewhat common, so it seems better not to break it.

GuillaumeGomez commented:
Oh I see, good point!

nikomatsakis commented:
Nominating for @rust-lang/lang meeting to draw attention to the PR.

nikomatsakis commented:
I'd like to see crater results, but I think this is reasonable. It'd be nice to have a survey of how other markdown systems approach this though.

Manishearth commented:
Also would like to see crater results, but I'm fine with this otherwise

@rfcbot approve

casey commented:
I did a mini survey of other markdown systems.

- **Github**: Splits on `,`, ` `, and `\t`.
- **Gitlab**: Doesn't split on anything. Entire info string is used verbatim, so `rust,foo` and `rust foo` disable syntax highlighting.
- **Misune**, a popular python markdown library: Splits on `,`, all `,` separated items are added to the class list in the generated HTML. Everything after the first space or tab is ignored, so `rust foo` has `class=language-rust`, and ignores `foo`.
- **Remark**, a popular JS markdown library: Splits on `,` and adds `,` separated items to class list in HTML. `rust foo` actually puts the text `foo` *into* the code block. `<code class="rust">foo</code>`.

All four treat characters other than `,`, ` `, `\t` the same as alphanumeric, i.e. the don't split, and are included in the parsed language string.

It seems like there's little agreement 😅 If there are other systems, that I should take a look at, let me know.

Do I need to do anything to trigger a crater run?

jyn514 commented:
I'm not sure if crater would tell us anything meaningful here, actually. Rustdoc doesn't warn or error when it runs into a language string it doesn't understand, so even if there were a change in behavior we wouldn't see it on crater.

casey commented:
We could see test failures if attributes stop being recognized. For example, this would currently get the `should_panic` attribute, but after the change wouldn't, and would fail:

````rust
```rust&should_panic
panic!();
```
````

It would be ideal to see if the number of tests that ran changed. If they stayed the same and nothing failed, we could be confident that there were no breakages.

Manishearth commented:
@jyn514 correct, but we can push a change that makes this a hard error and then crater it

jyn514 commented:
Got it, thanks.

@casey can you change the PR to give an error/give an assertion failure if the new and old parsing don't match? You'll have to temporarily disable the tests. Then I can start a crater run to see how often this happens in practice.

casey commented:
I think this is ready for a crater run. I modified it to parse lang strings twice, once with the new splitting scheme, and once with the old, and panic if the parsed `LangString` objects don't match.

jyn514 commented on src/test/rustdoc/issue-32556.rs:
````diff
@@ -1,5 +1,5 @@
 /// Blah blah blah
-/// ```ignore (testing rustdoc's handling of ignore)
+/// ```ignore (test rustdocs handling of ignor)
````
I don't think you meant to change this.
```suggestion
/// ```ignore (testing rustdoc's handling of ignore)
```

jyn514 commented on compiler/rustc_error_codes/src/error_codes/E0761.md:
````diff
@@ -2,7 +2,7 @@ Multiple candidate files were found for an out-of-line module.
 Erroneous code example:
 
-```ignore (multiple source files required for compile_fail)
+```ignore (Multiple source files are required for compile fail.)
````
```suggestion
```ignore (Multiple source files are required for compile_fail.)
```

jyn514 reviewed (commented):
I'm building it locally to make sure it gives an assertion failure. I have a few minor nits in the meantime.

jyn514 commented on src/librustdoc/html/markdown.rs:
```diff
@@ -797,11 +797,50 @@ impl LangString {
+            true,
+        );
+
+        assert_eq!(
+            new, old,
+            "New token split differs from old split.\nNew: {:?}\nOld: {:?}",
+            new, old
+        );
```
This prints the assertion failure twice:

```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `LangString { original: "{.no_run .example}", should_panic: false, no_run: false, ignore: None, rust: false, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`,
 right: `LangString { original: "{.no_run .example}", should_panic: false, no_run: true, ignore: None, rust: true, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`: New token split differs from old split.
New: LangString { original: "{.no_run .example}", should_panic: false, no_run: false, ignore: None, rust: false, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }
Old: LangString { original: "{.no_run .example}", should_panic: false, no_run: true, ignore: None, rust: true, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }', src/librustdoc/html/markdown.rs:829:9
```

casey replied:
Fixed

casey replied:
Fixed

casey commented on src/librustdoc/html/markdown/tests.rs:
```diff
@@ -97,12 +97,10 @@ fn test_lang_string_parse() {
-    t("{.example .rust}", false, false, Ignore::None, true, false, false, false, v(), None);
-    t("{.test_harness .rust}", false, false, Ignore::None, true, true, false, false, v(), None);
     t("text, no_run", false, true, Ignore::None, false, false, false, false, v(), None);
     t("text,no_run", false, true, Ignore::None, false, false, false, false, v(), None);
+    t("text,no_run,", false, true, Ignore::None, false, false, false, false, v(), None);
+    t("text,no_run, ", false, true, Ignore::None, false, false, false, false, v(), None);
```
They're very slightly different, the latter two have a trailing comma, and a trailing comma and space.

casey commented on src/librustdoc/html/markdown/tests.rs:
```diff
@@ -97,12 +97,10 @@ fn test_lang_string_parse() {
     t("test_harness", false, false, Ignore::None, true, true, false, false, v(), None);
     t("compile_fail", false, true, Ignore::None, true, false, true, false, v(), None);
     t("allow_fail", false, false, Ignore::None, true, false, false, true, v(), None);
-    t("{.no_run .example}", false, true, Ignore::None, true, false, false, false, v(), None);
-    t("{.sh .should_panic}", true, false, Ignore::None, false, false, false, false, v(), None);
-    t("{.example .rust}", false, false, Ignore::None, true, false, false, false, v(), None);
-    t("{.test_harness .rust}", false, false, Ignore::None, true, true, false, false, v(), None);
```
Good catch, I'll add them back with the correct syntax.

casey replied:
Fixed

casey commented:
Those two comment changes you noticed were actually necessary to get the parser to return the same result for those two lines. Using the old splitter `foo)` would split as `foo`, but the new one splits it as `foo)`. This led the new splitter to ignore those tokens as invalid, which changed the way the lang string was parsed. I think the difference in parsing actually doesn't make a difference, it only changes whether or not the `rust` bool is set on the lang string, which doesn't make a difference if it's already `ignore`ed.

Other than that, hopefully I addressed all your comments.

nikomatsakis commented:
@rfcbot reviewed 

I would be happy to match github -- it sounds like the newly proposed behavior is "closer" to what other markdown processors do, in any case.

jyn514 commented:
@craterbot run mode=rustdoc

jyn514 commented:
@bors try

bors commented:
:hourglass: Trying commit 46b8e7e537d8b9feda411e96a6e760155e1bc94b with merge 2880e4bd9459f3eab950f186eff1ff2d475dd30a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"46b8e7e537d8b9feda411e96a6e760155e1bc94b","merge_sha":"2880e4bd9459f3eab950f186eff1ff2d475dd30a"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1342978440)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1342978440","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job `dist-x86_64-linux` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/1342825058) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/2880e4bd9459f3eab950f186eff1ff2d475dd30a/checks/1342825058/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/2880e4bd9459f3eab950f186eff1ff2d475dd30a/checks/1342825058/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
 Documenting rustc_typeck v0.0.0 (/checkout/compiler/rustc_typeck)
    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
 Documenting rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `LangString { original: "ignore(cross-crate-imports)", should_panic: false, no_run: false, ignore: None, rust: false, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`,
 right: `LangString { original: "ignore(cross-crate-imports)", should_panic: false, no_run: false, ignore: All, rust: true, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`: New token split differs from old token split', src/librustdoc/html/markdown.rs:829:9

error: internal compiler error: unexpected panic


error: Unrecognized option: 'crate-version'
error: could not document `rustc_mir`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2018 --crate-type lib --crate-name rustc_mir compiler/rustc_mir/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/doc --error-format=json --json=diagnostic-rendered-ansi -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/release/deps --extern either=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libeither-0886f79aa9644b59.rmeta --extern itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libitertools-15ba3a79d5ec8f6f.rmeta --extern polonius_engine=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libpolonius_engine-9a99b9287362070b.rmeta --extern regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libregex-32ea2f287d81478c.rmeta --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-2bc88092b3516846.rmeta --extern rustc_ast=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_ast-9927a9ba1f5c66b9.rmeta --extern rustc_attr=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_attr-05146f72b221865d.rmeta --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-fd9a37813ceff842.rmeta --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-77a71cbc940ce463.rmeta --extern rustc_graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_graphviz-e79b54e3fcce812e.rmeta --extern rustc_hir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_hir-0552a4384760d9f0.rmeta --extern rustc_index=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_index-78f7d4c857d61346.rmeta --extern rustc_infer=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_infer-84826c8a1e892115.rmeta --extern rustc_lexer=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_lexer-d9c58c19eb967beb.rmeta --extern rustc_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/release/deps/librustc_macros-526b2b3435deae8e.so --extern rustc_middle=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_middle-71b5f6eea2ba7152.rmeta --extern rustc_serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_serialize-ba853f2a3b898296.rmeta --extern rustc_session=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_session-4947eec4f6403f55.rmeta --extern rustc_span=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_span-055868e9e3b7af54.rmeta --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-33cb9f703ec451d2.rmeta --extern rustc_trait_selection=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_trait_selection-c124db26a8699570.rmeta --extern smallvec=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libsmallvec-31c79deb030c1dbf.rmeta --extern tracing=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libtracing-e9500837597e63e7.rmeta -Dwarnings -Winvalid_codeblock_attributes --crate-version '1.49.0-nightly
  (2880e4bd9
  2020-11-02)' --document-private-items --enable-index-page -Zunstable-options` (exit code: 1)
error: build failed



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "doc" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "jemalloc llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "--no-deps" "-p" "rustc_parse_format" "-p" "rustc_session" "-p" "rustc_ast_passes" "-p" "rustc_data_structures" "-p" "rustc_parse" "-p" "rustc_query_system" "-p" "rustc_apfloat" "-p" "rustc_ty" "-p" "rustc_attr" "-p" "rustc_index" "-p" "rustc_symbol_mangling" "-p" "rustc_span" "-p" "rustc_infer" "-p" "rustc_errors" "-p" "rustc_codegen_ssa" "-p" "rustc_fs_util" "-p" "rustc_save_analysis" "-p" "rustc_mir" "-p" "rustc_arena" "-p" "rustc_serialize" "-p" "rustc_feature" "-p" "rustc_middle" "-p" "rustc_traits" "-p" "rustc_interface" "-p" "rustc_incremental" "-p" "rustc_lexer" "-p" "rustc_resolve" "-p" "rustc_metadata" "-p" "rustc_lint" "-p" "rustc_typeck" "-p" "rustc_ast" "-p" "rustc_passes" "-p" "rustc_expand" "-p" "rustc_macros" "-p" "rustc_error_codes" "-p" "rustc_hir_pretty" "-p" "rustc_codegen_llvm" "-p" "rustc_privacy" "-p" "rustc_llvm" "-p" "rustc_ast_pretty" "-p" "rustc_trait_selection" "-p" "rustc_lint_defs" "-p" "rustc_ast_lowering" "-p" "rustc_driver" "-p" "rustc_mir_build" "-p" "rustc_graphviz" "-p" "rustc_builtin_macros" "-p" "rustc_plugin_impl" "-p" "rustc_target" "-p" "rustc_hir"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap dist --host x86_64-unknown-linux-gnu --target x86_64-unknown-linux-gnu --include-default-paths src/tools/build-manifest
Build completed unsuccessfully in 0:25:39
Build completed unsuccessfully in 0:25:39
== clock drift check ==
  local time: Mon Nov  2 15:54:05 UTC 2020
  network time: Sun, 01 Nov 2020 19:11:02 GMT
== end clock drift check ==
##[error]Process completed with exit code 1.
Terminate orphan process: pid (3040) (python)
```

</details><p></p>

[I'm a bot](https://github.com/rust-lang/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-lang/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

jyn514 commented:
```
     Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
 Documenting rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `LangString { original: "ignore(cross-crate-imports)", should_panic: false, no_run: false, ignore: None, rust: false, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`,
 right: `LangString { original: "ignore(cross-crate-imports)", should_panic: false, no_run: false, ignore: All, rust: true, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`: New token split differs from old token split', src/librustdoc/html/markdown.rs:829:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
Hmm, maybe rustdoc should also split on parentheses? I'd also be ok with just changing the compiler ignores and leaving rustdoc as-is, I expect we use `// ignore` more than most of the ecosystem.

casey commented:
> Hmm, maybe rustdoc should also split on parentheses?

I could also make it so that anything inside of matched parentheses is ignored, so it really functions as a comment, which seems to be the convention used in the rust codebase.

This is a little weird, and precludes using parentheses in attributes. The latter is unlikely, but I can imagine it happening if there is some kind of attribute that has a free text in it, e.g. `rust label=foo(bar)`, or if for some reason we introduced an attribute with a function-call-like syntax, e.g. `rust foo(1)`.

> I'd also be ok with just changing the compiler ignores and leaving rustdoc as-is, I expect we use // ignore more than most of the ecosystem.

I grepped, and it looks like the one that failed is the only such line that's likely to fail, and could be fixed by adding a space after ignore.

jyn514 commented:
@casey let's add a space for now then, and we can change the parsing if it turns out in the crater run that other people are using this.

casey commented:
@jyn514 Sounds good, just added spaces to the two failing langstrings.

jyn514 commented:
@bors try

bors commented:
:hourglass: Trying commit dcdf72e86fbf44de52140ac78904616274cc0b47 with merge 98096ec38c7951c6767a6ffc0523271fdc0043a1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dcdf72e86fbf44de52140ac78904616274cc0b47","merge_sha":"98096ec38c7951c6767a6ffc0523271fdc0043a1"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1364212927)
Build commit: 98096ec38c7951c6767a6ffc0523271fdc0043a1 (`98096ec38c7951c6767a6ffc0523271fdc0043a1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1364212927"},"merge_sha":"98096ec38c7951c6767a6ffc0523271fdc0043a1"} -->

jyn514 commented:
@craterbot run mode=rustdoc

:crossed_fingers:

jyn514 commented:
Looks like most of the regressions come from parsing `.ignore` differently from `ignore` now.

casey commented:
This is my first time reading a crater report. Does it deduplicate failures in dependencies? I.e. if two root crates fail because of the same dependency, does it report two failures or one?

I downloaded and grepped through the regressed crate build failures, and there are 1261 failures, and all but 176 are for the string `.ignore`.

There are a small number of unique cases. Here they are, along with the number of times they appear:

```
1085 .ignore
  94 {rust}
  20 {rust,ignore}
  14 {.should_panic}
   9 {.rust}
   5 rust{compile_fail}
   5 .rust
   4 rust#ignore
   3 [no_run]
   3 !ignore
   2 rust.ignore
   2 ignore{.should_panic}
   2 # compile_fail
   1 ¨
   1 {.ignore .rust}
   1 should_panic(expected = \"Method sequence violation\")
   1 rust{.should_panic}
   1 rust{.ignore}
   1 rust{.example}
   1 rust\u{328}
   1 rust,should_panic=\"Assert CLI failure\"
   1 rust(ignore)
   1 no_run(
   1 ignore=
   1 ;
   1 //!
```

I have no idea how to put these numbers in perspective, so very curious what others make of this.

jyn514 commented on compiler/rustc_error_codes/src/error_codes/E0761.md:
````diff
@@ -2,7 +2,7 @@ Multiple candidate files were found for an out-of-line module.
 Erroneous code example:
 
-```ignore (multiple source files required for compile_fail)
+```ignore (Multiple source files are required for compile_fail )
````
This can be reverted, I think.

casey replied:
Reverted.

jyn514 commented on src/librustdoc/html/markdown.rs:
```diff
@@ -797,11 +797,46 @@ impl LangString {
         enable_per_target_ignores: bool,
         extra: Option<&ExtraInfo<'_, '_>>,
+    ) -> LangString {
+        let new = Self::parse_inner(
+            string,
+            allow_error_code_check,
+            enable_per_target_ignores,
+            extra,
+            false,
+        );
+
+        let old = Self::parse_inner(
+            string,
+            allow_error_code_check,
+            enable_per_target_ignores,
+            extra,
+            true,
+        );
+
+        assert_eq!(new, old, "New token split differs from old token split");
+
+        new
+    }
+
+    fn parse_inner(
```
I think most of this can been reverted now that we've done a crater run.

casey replied:
Done!

jyn514 commented on src/test/rustdoc/issue-32556.rs:
````diff
@@ -1,5 +1,5 @@
 /// Blah blah blah
-/// ```ignore (testing rustdoc's handling of ignore)
+/// ```ignore (testing rustdoc's handling of ignore )
````
This can be reverted.

casey replied:
Reverted.

jyn514 reviewed (commented):
I think this is still waiting on some more people from lang team/rustdoc to sign off. Is anyone concerned by the regressions in crater or do they all look niche enough not to worry about? We could fix almost all of them by also splitting on `{}.`, I'd be fine with that.

jyn514 commented:
ping @GuillaumeGomez @ollie27 @pnkfelix @withoutboats - this is waiting on your review: https://github.com/rust-lang/rust/pull/78429#issuecomment-716942661

GuillaumeGomez commented:
Sorry, completely forgot about it. Code looks good to me!

bors commented:
:umbrella: The latest upstream changes (presumably #80024) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

bors commented:
:umbrella: The latest upstream changes (presumably #80206) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1604268168) [(plain)](https://github.com/rust-lang/rust/commit/3805c7d691642b4af036c310e8a40e50a3157503/checks/1604268168/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9000/11197
.................................................................................................... 9100/11197
.......................................................................................i......i..... 9200/11197
.................................................................................................... 9300/11197
..........................iiiiii..iiiiii.i.......................................................... 9400/11197
.................................................................................................... 9600/11197
.................................................................................................... 9700/11197
.................................................................................................... 9800/11197
.................................................................................................... 9900/11197
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 27 tests
iiiiiiiiiiiiiiiiiiiiiiiiiii

 finished in 0.074 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i....i.i.ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.30s

 finished in 2.373 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
running 50 tests
................................................F.
failures:

---- html::markdown::tests::test_lang_string_parse stdout ----
thread 'html::markdown::tests::test_lang_string_parse' panicked at 'assertion failed: `(left == right)`
  left: `LangString { original: ".no_run,.example}", should_panic: false, no_run: false, ignore: None, rust: false, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`,
 right: `LangString { original: ".no_run,.example}", should_panic: false, no_run: true, ignore: None, rust: true, test_harness: false, compile_fail: false, error_codes: [], allow_fail: false, edition: None }`', src/librustdoc/html/markdown/tests.rs:56:9


failures:
    html::markdown::tests::test_lang_string_parse
    html::markdown::tests::test_lang_string_parse

test result: FAILED. 49 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

error: test failed, to rerun pass '-p rustdoc --lib'


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/rustdoc/Cargo.toml" "-p" "rustdoc:0.0.0" "--" "--quiet"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:30:59
```

</details>

casey commented:
I can't figure out how to run the tests in `src/librustdoc/html/markdown/tests.rs`.

I thought that `x.py src/test/rustdoc` would run them, but I inserted intentional errors into `src/librustdoc/html/markdown/tests.rs`, and it looks like it doesn't.

`cargo test -p rustdoc` and `cargo test` while in `src/librustdoc` don't work either:

```
error[E0463]: can't find crate for `lazy_static`
  --> src/librustdoc/lib.rs:23:1
   |
23 | extern crate lazy_static;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error
```

Is there another way to run these tests?

jyn514 commented:
@casey try `x.py test --stage 0 src/librustdoc`.

casey commented:
@jyn514 Ahhh, thank you! Tests should be fixed.

jyn514 commented:
ping @ollie27 @pnkfelix @withoutboats - this is waiting on your review: https://github.com/rust-lang/rust/pull/78429#issuecomment-716942661. The crater run is here if you want to read it: https://github.com/rust-lang/rust/pull/78429#issuecomment-726485338

ollie27 commented:
Sorry for the slow reply.

Given the amount of breakage found by crater I do think we should continue to support the `{.should_panic}` like syntax by splitting on `{`, `}` and `.` at least. Although, does anyone know where that syntax originally came from?

jyn514 commented:
> Although, does anyone know where that syntax originally came from?

It comes from pandoc: https://pandoc.org/MANUAL.html#extension-fenced_code_attributes. The first commit adding it was fefdb63c4c2b078c43836e2ae9d7ffcaeec32890.

@casey can you change this to split on `{}.` as well? Thank you for sticking with this :heart:

casey commented:
> It comes from pandoc: https://pandoc.org/MANUAL.html#extension-fenced_code_attributes. The first commit adding it was [fefdb63](https://github.com/rust-lang/rust/commit/fefdb63c4c2b078c43836e2ae9d7ffcaeec32890).

Nice sleuthing!

> @casey can you change this to split on `{}.` as well?

I can definitely do that, although what do you think about doing the minimum possible to confirm to the pandoc syntax.

I could:

1. Remove `{` and `}` if they are the first and last character, respectively, of the langstring.
2. Remove a single leading `.` if present, from each token.

This would allow `{`, and `}` to be used freely within lang strings. (Assuming they don't also start with a `{`, in which case a trailing `}` would also be stripped.)

And `.` could also be used within lang string tokens and values, so `foo.bar=baz` or `baz=foo.bar` would both be possible, if so desired, in future extensions.

> Thank you for sticking with this ❤️

You bet! Thank you for your patience and review.

casey commented:
Additionally, I could also add warnings when the special behavior of `{}.` was triggered, so that it could be removed some time in the future.

jyn514 commented:
Personally, I don't think there's a need to remove `{}.`, there's no proposed use for them and I'd prefer not to have things change meaning.

I don't have a preference for whether to always split on `{}.` or the most target approach you suggested. I slightly favor the more target approach just because it gives rustdoc more flexibility.

casey commented:
> Personally, I don't think there's a need to remove `{}.`, there's no proposed use for them and I'd prefer not to have things change meaning.

I definitely wouldn't propose changing the meaning, but one benefit of just adding a warning would be discouraging people from inadvertently relying on the special behavior of `{}.`. It might be annoying if a the existing code started throwing warnings though, so I think whether or not a warning is a good idea would also strongly depend on whether or not the existing uses would get fixed eventually, or would just throw warnings forever.

> I don't have a preference for whether to always split on `{}.` or the most targeted approach you suggested. I slightly favor the more targeted approach just because it gives rustdoc more flexibility to add things later.

Sounds good, I'll update this PR with the targeted approach.

casey commented:
@jyn514 I just pushed a change that strips a surrounding `{}` from the langstring, and a leading `.` from each token, along with some test cases.

jyn514 commented:
Thanks, looks great! @ollie27 can you take another look? I think we'll be ready to go after that :)

@casey in the meantime do you mind squashing the commits to something fewer than 24?

casey commented:
> @casey in the meantime do you mind squashing the commits to something fewer than 24?

Yeah, 24 is a bit excessive for this change 😅 I squashed to one commit, and improved the commit message.

dtolnay commented:
I've approved on behalf of withoutboats, as per https://github.com/rust-lang/team/pull/526.

jyn514 commented:
@bors r+

bors commented:
:pushpin: Commit 66f4883308d999c8b405fdfd442562b8600a462d has been approved by `jyn514`

<!-- @bors r=jyn514 66f4883308d999c8b405fdfd442562b8600a462d -->
<!-- homu: {"type":"Approved","sha":"66f4883308d999c8b405fdfd442562b8600a462d","approver":"jyn514"} -->

bors commented:
:hourglass: Testing commit 66f4883308d999c8b405fdfd442562b8600a462d with merge d95d30486180387a875b14633aea4e4dd8f960da...
<!-- homu: {"type":"BuildStarted","head_sha":"66f4883308d999c8b405fdfd442562b8600a462d","merge_sha":"d95d30486180387a875b14633aea4e4dd8f960da"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1984499589)
Approved by: jyn514
Pushing d95d30486180387a875b14633aea4e4dd8f960da to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jyn514","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1984499589"},"merge_sha":"d95d30486180387a875b14633aea4e4dd8f960da"} -->

casey commented:
Awesome! Thank you @jyn514 for helping me get this through!

> As the automated representative of the governance process, I would like to thank the author for their work and everyone else who contributed.

I love how polite rfcbot is 😊

jyn514 commented:
You're very welcome! Thanks for sticking with it so long :)
